### PR TITLE
Add basic auth for Registry

### PR DESCRIPTION
**Pull request by deler-aziz**
Use a secret volume mount for htpasswd (basicAuth)

### CODE REVIEW

**Comment by nurimansyah89** [comment]
Hi @deler-aziz , nice addition, but I think you should add some configuration in the registry config too as well based on this:
https://docs.docker.com/registry/configuration/#auth

cmiiw 👍 👍 

---
The easiest solution is to override with environment variable, like so:
```
REGISTRY_AUTH_HTPASSWD_REALM: basic-realm
REGISTRY_AUTH_HTPASSWD_PATH: /host/path (Maybe we can substitute with the helm values as well)
```

> Or, just make some clear description about how to enable the htpasswd authentication by passing additional extra environments or make it static inside the `registry-deployment.yaml` file.

Wdyt?

**Comment by deler-aziz** [comment]
> Hi @deler-aziz , nice addition, but I think you should add some configuration in the registry config too as well based on this: https://docs.docker.com/registry/configuration/#auth
> 
> cmiiw 👍 👍
> 
> The easiest solution is to override with environment variable, like so:
> 
> ```
> REGISTRY_AUTH_HTPASSWD_REALM: basic-realm
> REGISTRY_AUTH_HTPASSWD_PATH: /host/path (Maybe we can substitute with the helm values as well)
> ```
> 
> > Or, just make some clear description about how to enable the htpasswd authentication by passing additional extra environments or make it static inside the `registry-deployment.yaml` file.
> 
> Wdyt?

Hi @nurimansyah89 
I agree, sounds like a good suggestion. I will push a new commit  👍

File: charts/docker-registry-ui/values.yaml [hunk 1]
```diff
@@ -170,6 +170,16 @@ registry:
...
   # Extra Environmental Variables for Registry
   extraEnv: []
+  # Add Basic Auth to Registry.
+  basicAuth:
+    # Enable basic auth for Registry.
+    enabled: false
+    # Basic auth realm.
+    realm: Docker registry
+    # Full path for htpasswd file. Note that filename should match the secret key.
+    htpasswdPath: /auth/htpasswd
```
Comment (Joxit) [inline charts/docker-registry-ui/values.yaml]: Move to `/etc/docker/registry/htpasswd` or `/etc/docker/registry/auth/htpasswd` ?

File: charts/docker-registry-ui/README.md [hunk 2]
```diff
@@ -121,6 +121,10 @@ helm upgrade --install docker-registry-ui joxit/docker-registry-ui
 | `registry.annotations` | `{}` | Annotations to apply to the registry server deployment. |
 | `registry.additionalSpec` | `{}` | Optional YAML string that will be appended to the deployment spec. |
 | `registry.extraEnv` | `[]` | Extra Environmental Variables for Registry |
+| `registry.basicAuth.enabled` | `false` | Enable basic auth for Registry.|
+| `registry.basicAuth.realm` | `Docker registry` | Basic auth realm.|
+| `registry.basicAuth.htpasswdPath` | `/auth/htpasswd` | Full path for htpasswd file. Note that filename should match the secret key.|
+| `registry.basicAuth.secretName` | `` | Basic Auth (htpasswd) Secret Name for Registry.|
```
Comment (Joxit) [inline charts/docker-registry-ui/README.md]: The README is managed by the CI and in house documentation generator

https://github.com/Joxit/helm-charts/blob/772be0269aa8d1b167a5b1190665a36916cf8692/.github/workflows/release.yaml#L20-L22

File: charts/docker-registry-ui/templates/registry-deployment.yaml [hunk 3]
```diff
@@ -51,6 +59,11 @@ spec:
           volumeMounts:
             - mountPath: /var/lib/registry
               name: data
+            {{- if .Values.registry.basicAuth.enabled }}
+            - name: htpasswd
+              mountPath: {{ if .Values.registry.basicAuth.htpasswdPath }}{{ dir .Values.registry.basicAuth.htpasswdPath }}{{ else }}{{ "/auth" }}{{ end }}
```
Comment (Joxit) [inline charts/docker-registry-ui/templates/registry-deployment.yaml]: The default value is already in `values.yaml` why should we test it again ?

File: charts/docker-registry-ui/templates/registry-deployment.yaml [hunk 4]
```diff
@@ -40,6 +40,14 @@ spec:
             - name: REGISTRY_STORAGE_DELETE_ENABLED
               value: 'true'
             {{- end }}
+            {{- if .Values.registry.basicAuth.enabled }}
+            - name: REGISTRY_AUTH
+              value: htpasswd
+            - name: REGISTRY_AUTH_HTPASSWD_REALM
+              value: {{ .Values.registry.basicAuth.realm | default "Docker registry"}}
+            - name: REGISTRY_AUTH_HTPASSWD_PATH
+              value: {{ .Values.registry.basicAuth.htpasswdPath | default "/auth/htpasswd" }}
```
Comment (Joxit) [inline charts/docker-registry-ui/templates/registry-deployment.yaml]: The default value is already in `values.yaml` why should add it again ?

**Review by Joxit — COMMENTED** [review]
Hi there, thank you for using my project and your contribution :smile: 

Can I have your feedback on the current configuration. Since you are adding basic auth, maybe one day someone will add token auth too.
In that case, in your opinion, what's the best configuration ? `registry.basicAuth.*` and `registry.tokenAuth.*` ? or `registry.auth.basic.*` and `registry.auth.token.*` ?

**Comment by nurimansyah89** [comment]
> Hi there, thank you for using my project and your contribution 😄
> 
> Can I have your feedback on the current configuration. Since you are adding basic auth, maybe one day someone will add token auth too. In that case, in your opinion, what's the best configuration ? `registry.basicAuth.*` and `registry.tokenAuth.*` ? or `registry.auth.basic.*` and `registry.auth.token.*` ?

Sorry for slow response, being hectic on office 😬 

I think we should do a better structure, like:
```
...
registry:
  auth:
    basic:
       ... # All the basic auth related
    custom:
       ... # Any custom auth
...
```

This just my opinion for better readability for using the values. The trade off this structure was it has a multilevel nesting.

Wdyt @Joxit ?

File: charts/docker-registry-ui/values.yaml (hunk 1 above)
Comment (nurimansyah89) [reply charts/docker-registry-ui/values.yaml]: sounds good to me, for making it more clean in one place 👍

File: charts/docker-registry-ui/templates/registry-deployment.yaml (hunk 3 above)
Comment (nurimansyah89) [reply charts/docker-registry-ui/templates/registry-deployment.yaml]: agree with this, except if the value in `values.yaml` are filled with `~`, we should declare default value inside the template.

**Comment by deler-aziz** [comment]
Hi @Joxit and @nurimansyah89 

Sorry for the slow response. You have left good feedback thank you for that. 
This is the new auth structure:
```
...
registry:
  auth:
    basic:
      enabled: false
      realm: Docker registry
      htpasswdPath: /etc/docker/registry/auth/htpasswd
      secretName: ''
...
```
Also, I have done some more changes, please check it and if you have any more feedback I will be glad to get it.

**Comment by nurimansyah89** [comment]
> Hi @Joxit and @nurimansyah89
> 
> Sorry for the slow response. You have left good feedback thank you for that. This is the new auth structure:
> 
> ```
> ...
> registry:
>   auth:
>     basic:
>       enabled: false
>       realm: Docker registry
>       htpasswdPath: /etc/docker/registry/auth/htpasswd
>       secretName: ''
> ...
> ```
> 
> Also, I have done some more changes, please check it and if you have any more feedback I will be glad to get it.

Looks good to me. 👍 

Just wait for @Joxit to review them and get merged. Nice work @deler-aziz 👍

**Review by Joxit — APPROVED** [review]

**Comment by Joxit** [comment]
Hi there, thank the both of you for your contribution and insights :smile:

**Pull request merged**